File: platform/macos/Bridge/NativeConnectionDocument.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Foundation
import TidyVNC

extension tidyvnc_document_info: ABIValue {}
extension tidyvnc_document_entry: ABIValue {}

public enum NativeDocumentProblem: UInt32, Sendable {
  case empty = 1, invalidHeader, nullByte, lineTooLong, invalidAssignment
  case invalidEscape, tooLarge, tooManyEntries, invalidExportName
  case invalidText, invalidIndex
  case invalidValue, unavailable
}
public struct NativeDocumentFailure: Error, Sendable, Equatable, CustomStringConvertible {
  public let problem: NativeDocumentProblem
  public let line: UInt32
  public var description: String {
    let message: String
    switch problem {
    case .empty: message = "The connection file is empty."
    case .invalidHeader: message = "The connection file header is unsupported."
    case .nullByte: message = "The connection file contains a null byte."
    case .lineTooLong: message = "A connection-file line exceeds its byte limit."
    case .invalidAssignment: message = "A connection-file assignment is malformed."
    case .invalidEscape: message = "A connection-file value contains an invalid escape."
    case .tooLarge: message = "The connection file exceeds its size limit."
    case .tooManyEntries: message = "The connection file contains too many assignments."
    case .invalidExportName: message = "A field cannot be exported to a connection file."
    case .invalidText: message = "The connection file is not valid UTF-8."
    case .invalidIndex: message = "The connection-file entry does not exist."
    case .invalidValue: message = "A connection-file option has an invalid value."
    case .unavailable: message = "A connection-file option is unavailable in this build."
    }
    return line == 0 ? message : "Line \(line): \(message)"
  }
}
@discardableResult private func documentCall(allowing: Set<NativeStatus> = [.ok], _ body: (UnsafeMutablePointer<tidyvnc_error>) -> UInt32) throws -> NativeStatus {
  do { return try checked(allowing: allowing, body) }
  catch let error as NativeError {
    if error.domain == UInt32(TIDYVNC_DOMAIN_DOCUMENT), let problem = NativeDocumentProblem(rawValue: error.detail & 255) {
      throw NativeDocumentFailure(problem: problem, line: error.detail >> 8)
    }
    throw error
  }
}
private func documentText<T>(_ value: T, line: UInt32) throws -> String {
  try withUnsafeBytes(of: value) { bytes in
    guard let result = String(bytes: bytes.prefix(while: { $0 != 0 }), encoding: .utf8) else {
      throw NativeDocumentFailure(problem: .invalidText, line: line)
    }
    return result
  }
}
public struct NativeDocumentAssignment: Sendable, Equatable {
  public let name: String, value: String
  public init(_ name: String, _ value: String) { self.name = name; self.value = value }
}
public struct NativeDocumentEntry: Sendable, Equatable {
  public let name: String, encodedValue: String
  public let line: UInt32
}

// Immutable core handle and copied Swift metadata. The synchronized registry
// protects concurrent readers; no borrowed bytes escape a call. This parses
// syntax only: it does not apply settings, read files or authorize migration.
public final class NativeConnectionDocument: Sendable {
  public static let maximumBytes = 1_048_576
  public static let maximumEntries = 4096
  private let handle: NativeHandle
  public let isLegacy: Bool
  public let entries: [NativeDocumentEntry]

  public init(data: Data) throws {
    guard data.count <= Self.maximumBytes else { throw NativeDocumentFailure(problem: .tooLarge, line: 0) }
    // Explicitly reject invalid text, including in ignored/comment fields;
    // never silently turn unrepresentable bytes into replacement characters.
    guard String(data: data, encoding: .utf8) != nil else { throw NativeDocumentFailure(problem: .invalidText, line: 0) }
    var raw: UInt64 = 0
    _ = try data.withUnsafeBytes { bytes in
      try documentCall { tidyvnc_document_parse(tidyvnc_bytes(data: bytes.bindMemory(to: UInt8.self).baseAddress, length: UInt64(bytes.count)), &raw, $0) }
    }
    let owner = NativeHandle(adopting: raw)
    var info = abi(tidyvnc_document_info.self)
    try documentCall { tidyvnc_document_get(owner.raw, &info, $0) }
    guard info.count <= Self.maximumEntries, info.legacy_header <= 1 else { throw NativeError(.unsupported, "Unsupported connection-file metadata") }
    var copied: [NativeDocumentEntry] = []
    copied.reserveCapacity(Int(info.count))
    for index in 0..<info.count {
      var entry = abi(tidyvnc_document_entry.self)
      try documentCall { tidyvnc_document_entry_at(owner.raw, index, 0, &entry, $0) }
      copied.append(try NativeDocumentEntry(name: documentText(entry.name, line: entry.line),
        encodedValue: documentText(entry.value, line: entry.line), line: entry.line))
    }
    handle = owner; isLegacy = info.legacy_header != 0; entries = copied
  }
  public func decodedValue(at index: Int) throws -> String {
    guard entries.indices.contains(index) else { throw NativeDocumentFailure(problem: .invalidIndex, line: 0) }
    var value = abi(tidyvnc_document_entry.self)
    try documentCall { tidyvnc_document_entry_at(handle.raw, UInt32(index), 1, &value, $0) }
    return try documentText(value.value, line: value.line)
  }
  public func validatedOption(at index: Int) throws -> NativeDocumentAssignment? {
    guard entries.indices.contains(index) else { throw NativeDocumentFailure(problem: .invalidIndex, line: 0) }
    var value = abi(tidyvnc_document_entry.self)
    if try documentCall(allowing: [.ok,.noChange], { tidyvnc_document_option_at(handle.raw, UInt32(index), &value, $0) }) == .noChange { return nil }
    return try .init(documentText(value.name, line: value.line), documentText(value.value, line: value.line))
  }
  public static func serialize(_ fields: [NativeDocumentAssignment]) throws -> Data {
    guard fields.count <= maximumEntries else { throw NativeDocumentFailure(problem: .tooManyEntries, line: 0) }
    var storage: [UInt8] = []
    var offsets: [(Int, Int, Int, Int)] = []
    for field in fields {
      guard field.name.utf8.prefix(256).count <= 255, field.value.utf8.prefix(256).count <= 255 else {
        throw NativeDocumentFailure(problem: .lineTooLong, line: 0)
      }
      let name = storage.count; storage.append(contentsOf: field.name.utf8)
      let value = storage.count; storage.append(contentsOf: field.value.utf8)
      offsets.append((name, value - name, value, storage.count - value))
    }
    return try storage.withUnsafeBufferPointer { buffer in
      let values = offsets.map { name, nameLength, value, valueLength in
        tidyvnc_document_assignment(name: tidyvnc_bytes(data: buffer.baseAddress?.advanced(by: name), length: UInt64(nameLength)),
          value: tidyvnc_bytes(data: buffer.baseAddress?.advanced(by: value), length: UInt64(valueLength)))
      }
      return try values.withUnsafeBufferPointer { input in
        var length: UInt64 = 0
        try documentCall { tidyvnc_document_serialize(input.baseAddress, UInt32(input.count), tidyvnc_mutable_bytes(data: nil, length: 0), &length, $0) }
        guard length <= maximumBytes else { throw NativeDocumentFailure(problem: .tooLarge, line: 0) }
        var result = Data(count: Int(length))
        _ = try result.withUnsafeMutableBytes { output in
          try documentCall { tidyvnc_document_serialize(input.baseAddress, UInt32(input.count),
            tidyvnc_mutable_bytes(data: output.bindMemory(to: UInt8.self).baseAddress, length: UInt64(output.count)), &length, $0) }
        }
        guard length == result.count else { throw NativeError(.internalFailure, "Connection-file export size changed") }
        return result
      }
    }
  }
}

File: platform/windows/TidyVNC.Native/Bridge/NativeConnectionDocument.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using System.Text;
using TidyVNC.Native.Interop;

namespace TidyVNC.Native;

/// <summary>tidyvnc document error reasons (NativeDocumentProblem in Swift).</summary>
public enum NativeDocumentProblem : uint
{
    Empty = 1, InvalidHeader, NullByte, LineTooLong, InvalidAssignment, InvalidEscape, TooLarge, TooManyEntries, InvalidExportName,
    InvalidText, InvalidIndex, InvalidValue, Unavailable,
}

/// <summary>A connection-file failure with its one-based line (zero when no line applies). Never carries file text.</summary>
public sealed class NativeDocumentFailure(NativeDocumentProblem problem, uint line) : Exception(line == 0 ? $"Connection file: {problem}" : $"Connection file line {line}: {problem}")
{
    public NativeDocumentProblem Problem { get; } = problem;
    public uint Line { get; } = line;
}

public sealed record NativeDocumentAssignment(string Name, string Value);

public sealed record NativeDocumentEntry(string Name, string EncodedValue, uint Line);

/// <summary>
/// A parsed .tidyvnc/.tigervnc connection file (tidyvnc_document_*;
/// NativeConnectionDocument on macOS). Syntax only: nothing is applied, read
/// or authorized. Invalid UTF-8 is rejected, never replaced.
/// </summary>
public sealed unsafe class NativeConnectionDocument : IDisposable
{
    public const int MaximumBytes = 1_048_576;
    public const int MaximumEntries = 4096;
    private static readonly UTF8Encoding Strict = new(false, true);

    private readonly NativeHandle handle;

    public bool IsLegacy { get; }
    public IReadOnlyList<NativeDocumentEntry> Entries { get; }

    private static NativeStatus Call(uint code, tidyvnc_error* error, NativeStatus allowed = NativeStatus.Ok)
    {
        var status = (NativeStatus)code;
        if (status == NativeStatus.Ok || status == allowed) return status;
        if (error->domain == Tidyvnc.TIDYVNC_DOMAIN_DOCUMENT && Enum.IsDefined((NativeDocumentProblem)(error->detail & 0xff)))
            throw new NativeDocumentFailure((NativeDocumentProblem)(error->detail & 0xff), error->detail >> 8);
        throw new NativeError(*error);
    }

    private static string Text(byte* bytes, int capacity, uint line)
    {
        var length = 0;
        while (length < capacity && bytes[length] != 0) length++;
        try { return Strict.GetString(bytes, length); }
        catch (DecoderFallbackException) { throw new NativeDocumentFailure(NativeDocumentProblem.InvalidText, line); }
    }

    public NativeConnectionDocument(ReadOnlySpan<byte> data)
    {
        if (data.Length > MaximumBytes) throw new NativeDocumentFailure(NativeDocumentProblem.TooLarge, 0);
        try { _ = Strict.GetCharCount(data); }
        catch (DecoderFallbackException) { throw new NativeDocumentFailure(NativeDocumentProblem.InvalidText, 0); }
        var error = Abi.Init<tidyvnc_error>();
        ulong raw = 0;
        fixed (byte* bytes = data) Call(NativeMethods.tidyvnc_document_parse(AbiText.Span(bytes, data.Length), &raw, &error), &error);
        handle = NativeHandle.Adopt(raw);
        try
        {
            var info = Abi.Init<tidyvnc_document_info>();
            Call(NativeMethods.tidyvnc_document_get(handle.Raw, &info, &error), &error);
            if (info.count > MaximumEntries || info.legacy_header > 1) throw new NativeError(NativeStatus.Unsupported, "Unsupported connection-file metadata");
            var entries = new NativeDocumentEntry[info.count];
            for (uint index = 0; index < info.count; index++)
            {
                var entry = Abi.Init<tidyvnc_document_entry>();
                Call(NativeMethods.tidyvnc_document_entry_at(handle.Raw, index, 0, &entry, &error), &error);
                entries[index] = new NativeDocumentEntry(Text(entry.name, 255, entry.line), Text(entry.value, 256, entry.line), entry.line);
            }
            IsLegacy = info.legacy_header != 0;
            Entries = entries;
        }
        catch
        {
            handle.Dispose();
            throw;
        }
    }

    public string DecodedValue(int index)
    {
        if ((uint)index >= (uint)Entries.Count) throw new NativeDocumentFailure(NativeDocumentProblem.InvalidIndex, 0);
        var error = Abi.Init<tidyvnc_error>();
        var entry = Abi.Init<tidyvnc_document_entry>();
        Call(NativeMethods.tidyvnc_document_entry_at(handle.Raw, (uint)index, 1, &entry, &error), &error);
        return Text(entry.value, 256, entry.line);
    }

    /// <summary>The canonical option for a known field; null for unknown names.</summary>
    public NativeDocumentAssignment? ValidatedOption(int index)
    {
        if ((uint)index >= (uint)Entries.Count) throw new NativeDocumentFailure(NativeDocumentProblem.InvalidIndex, 0);
        var error = Abi.Init<tidyvnc_error>();
        var entry = Abi.Init<tidyvnc_document_entry>();
        if (Call(NativeMethods.tidyvnc_document_option_at(handle.Raw, (uint)index, &entry, &error), &error, NativeStatus.NoChange) == NativeStatus.NoChange)
            return null;
        return new NativeDocumentAssignment(Text(entry.name, 255, entry.line), Text(entry.value, 256, entry.line));
    }

    /// <summary>Canonical current header, names and LF endings; only the shared export catalog.</summary>
    public static byte[] Serialize(IReadOnlyList<NativeDocumentAssignment> fields)
    {
        if (fields.Count > MaximumEntries) throw new NativeDocumentFailure(NativeDocumentProblem.TooManyEntries, 0);
        var names = new byte[fields.Count][];
        var values = new byte[fields.Count][];
        for (var i = 0; i < fields.Count; i++)
        {
            names[i] = Encoding.UTF8.GetBytes(fields[i].Name);
            values[i] = Encoding.UTF8.GetBytes(fields[i].Value);
            if (names[i].Length > 255 || values[i].Length > 255) throw new NativeDocumentFailure(NativeDocumentProblem.LineTooLong, 0);
        }
        var handles = new List<System.Runtime.InteropServices.GCHandle>();
        try
        {
            var assignments = new tidyvnc_document_assignment[fields.Count];
            for (var i = 0; i < fields.Count; i++)
            {
                var name = System.Runtime.InteropServices.GCHandle.Alloc(names[i], System.Runtime.InteropServices.GCHandleType.Pinned);
                handles.Add(name);
                var value = System.Runtime.InteropServices.GCHandle.Alloc(values[i], System.Runtime.InteropServices.GCHandleType.Pinned);
                handles.Add(value);
                assignments[i] = new tidyvnc_document_assignment
                {
                    name = AbiText.Span((byte*)name.AddrOfPinnedObject(), names[i].Length),
                    value = AbiText.Span((byte*)value.AddrOfPinnedObject(), values[i].Length),
                };
            }
            var error = Abi.Init<tidyvnc_error>();
            ulong length = 0;
            fixed (tidyvnc_document_assignment* input = assignments)
            {
                Call(NativeMethods.tidyvnc_document_serialize(input, (uint)assignments.Length, default, &length, &error), &error);
                if (length > MaximumBytes) throw new NativeDocumentFailure(NativeDocumentProblem.TooLarge, 0);
                var output = new byte[length];
                fixed (byte* o = output)
                {
                    var written = length;
                    Call(NativeMethods.tidyvnc_document_serialize(input, (uint)assignments.Length,
                        new tidyvnc_mutable_bytes { data = o, length = (ulong)output.Length }, &written, &error), &error);
                    if (written != (ulong)output.Length) throw new NativeError(NativeStatus.InternalFailure, "Connection-file export size changed");
                }
                return output;
            }
        }
        finally
        {
            foreach (var pinned in handles) pinned.Free();
        }
    }

    public void Dispose() => handle.Dispose();
    public override string ToString() => "NativeConnectionDocument(<redacted>)";
}
